lightning/src/chain/channelmonitor.rs

```diff
@@ -2272,6 +2113,70 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
 	}
 }
 
+/// Persist is responsible for persisting channel monitors: this could mean
+/// writing once to disk, and/or uploading to one or more backup services.
+///
+/// Note that for every new monitor, you **must** persist the new ChannelMonitor
+/// to disk/backups. And, on every update, you **must** persist either the
+/// ChannelMonitorUpdate or the updated monitor itself. Otherwise, there is risk
+/// of situations such as revoking a transaction, then crashing before this
+/// revocation can be persisted, then unintentionally broadcasting a revoked
+/// transaction and losing money. This is a risk because previous channel states
+/// are toxic, so it's important that whatever channel state is persisted is
+/// kept up-to-date.
+///
+/// There are several different strategies for persistence, with different
+/// requirements. While not exhaustive, here are a few. A user could:
+///   * have multiple redundant backups where the user wants to ensure that
+///     every backup is always accessible. In this case, implementors of this
+///     trait need to ensure that every copy is updated at each step before
+///     moving forward, with none falling out of sync,
+///   * have backups where they're OK with relying on only M of N backups
+///     temporarily, in which case the user doesn't care about consensus between
+///     backups and only needs to ensure that they read from the backup(s) with
+///     the highest [`ChannelMonitorUpdate::update_id`].
```

Comment:
Honestly, if we keep filling every doc with twenty cases about monitor sync, we're gonna overload users everywhere. We need to centralize this documentation and describe several different approaches to both chain sync and monitor storage and then just link to it everywhere. It could be in a doc comment somewhere, or it could be a readme file. It may make sense as module-level documentation for `chain::channelmonitor`?

Reply:
I'm inclined to see us shift more towards using module-level documentation. I spelled out some reasoning in https://github.com/rust-bitcoin/rust-lightning/pull/649#discussion_r495113558. It also tends to be more discoverable if someone is browsing the docs.

Additionally, we can make those docs more approachable by breaking them up with headers, bullets, examples, links, etc., allowing users to scan and find broader topics instead of overwhelming them with large amounts of texts to wade through.

Reply:
Yes that's really something to think of as part of the dev onboarding.

I'm leaning toward independent readme files in a `doc/`. Like a BACKEND.md with examples, redirections towards module-level docs, implications.

Reply:
In general I like module-docs since they're staring you in the face when you read docs.rs and such, plus its easier to link to from other documentation. That said, for non-rust users module-level documentation may dissapear entirely by the time you see the C .h file, though, so we'd need some standard for copying it over if we go that route. Of course with .md docs we'd still need to figure out how to copy it over, so maybe its not really a major point. In any case, given the relative undocumented-ness of multi-monitor setups (and that we're tracking it at #604), I'd vote we not try to over-explain all the nuance in this PR.

Reply:
I agree there's a lot of nuance here (and agree about module-level docs being the way to go), so I removed the semi-half-assed backup consensus docs to push it off to whomever addresses #604.